apps/prairielearn/src/pages/instructorFileEditor/instructorFileEditor.js

```diff
@@ -466,616 +453,148 @@ async function writeEdit(fileEdit) {
   return fileID;
 }
 
-function saveAndSync(fileEdit, locals, callback) {
-  const options = {
-    course_id: locals.course.id,
-    user_id: locals.user.user_id,
-    authn_user_id: locals.authz_data.authn_user.user_id,
+async function saveAndSync(fileEdit, locals) {
+  const serverJob = await createServerJob({
+    courseId: locals.course.id,
+    userId: locals.user.user_id,
+    authnUserId: locals.authz_data.authn_user.user_id,
     type: 'sync',
     description: 'Save and sync an in-browser edit to a file',
-    courseDir: locals.course.path,
-  };
+  });
 
-  serverJobs.createJobSequence(options, (err, job_sequence_id) => {
-    // Return immediately if we fail to create a job sequence
-    if (ERR(err, callback)) return;
+  await serverJob.execute(async (job) => {
+    await updateJobSequenceId(fileEdit, serverJob.jobSequenceId);
 
-    let gitEnv = process.env;
+    const gitEnv = process.env;
     if (config.gitSshCommand != null) {
       gitEnv.GIT_SSH_COMMAND = config.gitSshCommand;
     }
 
-    let courseLock;
-    let jobSequenceHasFailed = false;
-    let startGitHash = null;
-    let endGitHash = null;
+    const lockName = `coursedir:${locals.course.path}`;
+    await namedLocks.tryWithLock(lockName, { timeout: 5000 }, async () => {
```

Comment:
Ah, there's something here we can bikeshed on a bit. In the old code where we'd use `tryLock` or `waitLock`, we'd manually check if `lock == null` to know if we acquired it. If we did not, we'd print a helpful error message like "Another user is already modifying this course".

In the new code with `doWithLock`/`tryWithLock`, however, the behavior is different:

- `doWithLock` will throw an error if the lock can't be acquired within the given timeout.
- `tryWithLock` will only execute the provided function if the lock can be acquired, but if it can't no error is thrown.

In both cases, we don't trivially get the opportunity to fail the job with a user-friendly error message ("Another user is already..."). My thinking with `tryWithLock` is that we could return a value from the inner function if the lock was acquired, check that value, and explicitly fail with a useful message if it couldn't be:

```ts
serverJob.execute(async (job) => {
  const didLock = await namedLocks.tryWithLock('name', {}, async () => {
    // Do things...
    return true;
  });
  if (!didLock) {
    job.fail('Another user is already modifying this course');
  }
});
```

(Note that I failed to do this in this PR).

What if instead of this, we deprecated `tryWithLock` entirely and let `doWithLock` take an optional fourth argument that is called on error?

```ts
serverJob.execute(async (job) => {
  await namedLocks.tryWithLock(
    'name',
    {},
    async () => {
      // Do things...
      return true;
    },
    () => {
      job.fail('Another user is already modifying this course');
    });
  });
});
```

Another option: we keep `tryWithLock` and mandate that it takes a fourth argument to handle the case where the lock can't be acquired? This would force us to be explicit about what happens when we don't acquire the lock.

Reply:
Can't `doWithLock` reject the promise and rely on `.catch()`? Not sure if you want to distinguish a "could-not-acquire-lock" error from an error triggered by the function, in that case this may make things less readable, but I'm ok with a callback on failing to acquire the lock, either on `doWithLock` optional or on `tryWithLock` required.

Reply:
We do indeed want to distinguish between those cases, yeah, and I don't think there's a straightforward way to do that with `catch`, unless we start throwing subclasses of errors that we can `instanceof` check?

Reply:
I like the idea of a mandatory fourth argument to handle the failed-to-acquire case. When reviewing https://github.com/PrairieLearn/PrairieLearn/pull/8111 I almost commented that the `didSync` logic was a bit sneaky and that it would be much nicer if there was some way to explicitly handle the failed-to-acquire case.

To bikeshed a bit further, an alternative to a fourth argument would be to pass the error handler in the `options` -- I think that the intent would be clearer if we could say `{onAcquireFail: () => {}}` rather than passing a mysterious fourth argument.

Also, should the error handler be `async` so that if we `await tryWithLock()` then we know we have successfully completed an async error handler if one was required?

Reply:
I like Matt's version better, as an option to `doWithLock`.

Reply:
I like the explicitness of that too. Taking that to its logical conclusion, what about an API like this?

```ts
await namedLocks.tryWithLock({
  name: `coursedir:${locals.course.path}`,
  timeout: 5000,
  onNotAcquired: () => {
    job.fail('Another user is already syncing or modifying this course.');
  },
  onAcquired: async () => {
    // Do stuff
  },
});
```

Of course, I can save this for another PR, as this would impact more callsites.

Reply:
I ended up implementing this over in #8111 as an optional property on the second argument. I'm going to make it mandatory when I do the larger refactor proposed in the above comment.

Reply:
Looks great!

Reply:
I updated this PR to use `onNotAcquired` in 823139e.